File: app/src/main/java/com/sjn/stamp/utils/TimeHelper.java
package com.sjn.stamp.utils;

import android.content.Context;
import android.content.res.Resources;

import com.sjn.stamp.R;

import org.joda.time.DateTime;
import org.joda.time.Duration;
import org.joda.time.LocalDate;
import org.joda.time.Minutes;
import org.joda.time.Months;
import org.joda.time.Seconds;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.ISODateTimeFormat;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class TimeHelper {

    public static String getDateText(Date date, Resources resources) {
        DateTime dateTime = TimeHelper.toDateTime(date).minusSeconds(20);
        DateTime now = TimeHelper.getJapanNow();
        Minutes minutes = Minutes.minutesBetween(dateTime, now);
        if (minutes.isLessThan(Minutes.minutes(1))) {
            return resources.getString(R.string.item_song_history_seconds_ago, Seconds.secondsBetween(dateTime, now).getSeconds());
        } else if (minutes.isLessThan(Minutes.minutes(60))) {
            return resources.getString(R.string.item_song_history_minutes_ago, Minutes.minutesBetween(dateTime, now).getMinutes());
        } else {
            return TimeHelper.formatMMDDHHMM(dateTime);
        }
    }

    static String getDateDiff(Context context, Date before) {
        return getDateDiff(context, before, getJapanNow().toDate());
    }

    private static String getDateDiff(Context context, Date before, Date after) {
        DateTime dt1 = toDateTime(before);
        DateTime dt2 = toDateTime(after);
        if (dt1.isAfter(dt2)) {
            dt1 = toDateTime(after);
            dt2 = toDateTime(before);
        }
        int months = Months.monthsBetween(dt1, dt2).getMonths();
        if (months > 12) {
            int years = months / 12;
            months = months - years * 12;
            return months == 0 ? context.getResources().getQuantityString(R.plurals.date_diff_years, years, years) :
                    context.getResources().getQuantityString(R.plurals.date_diff_years, years, years) + " " + context.getResources().getQuantityString(R.plurals.date_diff_months, months, months);
        } else if (months > 0) {
            return context.getResources().getQuantityString(R.plurals.date_diff_months, months, months);
        }
        Duration d = new Duration(dt1, dt2);
        if (d.getStandardDays() > 1) {
            int days = (int) d.getStandardDays();
            return context.getResources().getQuantityString(R.plurals.date_diff_days, days, days);
        }
        if (d.getStandardHours() >= 1) {
            int hours = (int) d.getStandardHours();
            int minutes = (int) d.getStandardMinutes() - hours * 60;
            return context.getResources().getQuantityString(R.plurals.date_diff_hours, hours, hours) + " " + context.getResources().getQuantityString(R.plurals.date_diff_minutes, minutes, minutes);
        } else {
            int minutes = (int) d.getStandardMinutes();
            return context.getResources().getQuantityString(R.plurals.date_diff_minutes, minutes, minutes);
        }
    }

    public static Date toDateOnly(Date date) {
        return toDateTime(date).withTimeAtStartOfDay().toDate();
    }

    public static LocalDate parse(String date) {
        return ISODateTimeFormat.localDateParser().parseLocalDate(date);
    }

    public static LocalDate getJapanToday() {
        return new LocalDate(new Date(System.currentTimeMillis()));
    }

    public static DateTime getJapanNow() {
        return new DateTime(new Date(System.currentTimeMillis()));
    }

    static String toRFC3339(long unixTime) {
        return ISODateTimeFormat.dateTime().print(new DateTime(new Date(unixTime * 1000L)));
    }

    public static String format(String dateTime) {
        return format(toDateTime(dateTime));
    }

    private static String format(DateTime dateTime) {
        return DateTimeFormat.mediumDateTime().print(dateTime);
    }

    public static String formatYYYYMMDDHHMMSS(Date date) {
        return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(date);
    }

    public static String formatMMDDHHMM(Date date) {
        return new SimpleDateFormat("MM/dd HH:mm", Locale.getDefault()).format(date);
    }

    public static String formatMMDDHHMM(DateTime date) {
        return new SimpleDateFormat("MM/dd HH:mm", Locale.getDefault()).format(date.toDate());
    }

    public static DateTime toDateTime(String rfc3339) {
        return ISODateTimeFormat.dateTime().parseDateTime(rfc3339);
    }

    public static DateTime toDateTime(Date date) {
        return new DateTime(date);
    }

    public static int getJapanMonth() {
        return getJapanNow().getMonthOfYear();
    }

    public static int getJapanYear() {
        return getJapanNow().getYear();
    }

    public static int getJapanDay() {
        return getJapanNow().getDayOfMonth();
    }

    public static LocalDate yearStart(int year) {
        return new LocalDate(year, 1, 1);
    }

    public static LocalDate yearEnd(int year) {
        return new LocalDate(year, 12, 31);
    }

    private static LocalDate yearMonthStart(int year, int month) {
        return new LocalDate(year, month, 1);
    }

    public static LocalDate yearMonthEnd(int year, int month) {
        return yearMonthStart(year, month).dayOfMonth().withMaximumValue();
    }
}

File: app/src/main/java/com/sjn/stamp/utils/TimeHelper.kt
package com.sjn.stamp.utils

import android.content.Context
import android.content.res.Resources
import com.sjn.stamp.R
import org.joda.time.*
import org.joda.time.format.DateTimeFormat
import org.joda.time.format.ISODateTimeFormat
import java.text.SimpleDateFormat
import java.util.*

@Suppress("unused")
object TimeHelper {

    val japanToday: LocalDate
        get() = LocalDate(Date(System.currentTimeMillis()))

    val japanNow: DateTime
        get() = DateTime(Date(System.currentTimeMillis()))

    val japanMonth: Int
        get() = japanNow.monthOfYear

    val japanYear: Int
        get() = japanNow.year

    val japanDay: Int
        get() = japanNow.dayOfMonth

    fun getDateText(date: Date, resources: Resources): String {
        val dateTime = TimeHelper.toDateTime(date).minusSeconds(20)
        val now = TimeHelper.japanNow
        val minutes = Minutes.minutesBetween(dateTime, now)
        return when {
            minutes.isLessThan(Minutes.minutes(1)) -> resources.getString(R.string.item_song_history_seconds_ago, Seconds.secondsBetween(dateTime, now).seconds)
            minutes.isLessThan(Minutes.minutes(60)) -> resources.getString(R.string.item_song_history_minutes_ago, Minutes.minutesBetween(dateTime, now).minutes)
            else -> TimeHelper.formatMMDDHHMM(dateTime)
        }
    }

    fun getDateDiff(context: Context, before: Date): String {
        return getDateDiff(context, before, japanNow.toDate())
    }

    fun getDateDiff(context: Context, before: Date, after: Date): String {
        var dt1 = toDateTime(before)
        var dt2 = toDateTime(after)
        if (dt1.isAfter(dt2)) {
            dt1 = toDateTime(after)
            dt2 = toDateTime(before)
        }
        var months = Months.monthsBetween(dt1, dt2).months
        if (months > 12) {
            val years = months / 12
            months -= years * 12
            return if (months == 0) context.resources.getQuantityString(R.plurals.date_diff_years, years, years) else context.resources.getQuantityString(R.plurals.date_diff_years, years, years) + " " + context.resources.getQuantityString(R.plurals.date_diff_months, months, months)
        } else if (months > 0) {
            return context.resources.getQuantityString(R.plurals.date_diff_months, months, months)
        }
        val d = Duration(dt1, dt2)
        if (d.standardDays > 1) {
            val days = d.standardDays.toInt()
            return context.resources.getQuantityString(R.plurals.date_diff_days, days, days)
        }
        return if (d.standardHours >= 1) {
            val hours = d.standardHours.toInt()
            val minutes = d.standardMinutes.toInt() - hours * 60
            context.resources.getQuantityString(R.plurals.date_diff_hours, hours, hours) + " " + context.resources.getQuantityString(R.plurals.date_diff_minutes, minutes, minutes)
        } else {
            val minutes = d.standardMinutes.toInt()
            context.resources.getQuantityString(R.plurals.date_diff_minutes, minutes, minutes)
        }
    }

    fun toDateOnly(date: Date): Date = toDateTime(date).withTimeAtStartOfDay().toDate()

    fun parse(date: String): LocalDate = ISODateTimeFormat.localDateParser().parseLocalDate(date)

    fun toRFC3339(unixTime: Long): String = ISODateTimeFormat.dateTime().print(DateTime(Date(unixTime * 1000L)))

    fun format(dateTime: String): String = format(toDateTime(dateTime))

    private fun format(dateTime: DateTime): String = DateTimeFormat.mediumDateTime().print(dateTime)

    fun formatYYYYMMDDHHMMSS(date: Date): String = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault()).format(date)

    fun formatMMDDHHMM(date: Date): String = SimpleDateFormat("MM/dd HH:mm", Locale.getDefault()).format(date)

    fun formatMMDDHHMM(date: DateTime): String = SimpleDateFormat("MM/dd HH:mm", Locale.getDefault()).format(date.toDate())

    fun toDateTime(rfc3339: String): DateTime = ISODateTimeFormat.dateTime().parseDateTime(rfc3339)

    fun toDateTime(date: Date): DateTime = DateTime(date)

    fun yearStart(year: Int): LocalDate = LocalDate(year, 1, 1)

    fun yearEnd(year: Int): LocalDate = LocalDate(year, 12, 31)

    private fun yearMonthStart(year: Int, month: Int): LocalDate = LocalDate(year, month, 1)

    fun yearMonthEnd(year: Int, month: Int): LocalDate = yearMonthStart(year, month).dayOfMonth().withMaximumValue()
}
